File: typescript/examples/agents/09-mcp-stdio.ts
// MCP Stdio Example — connect to a local MCP server and use its tools
//
// This example starts a local filesystem MCP server via stdio and lets
// the agent use it to list and read files.
//
// Prerequisites:
//   npm install @modelcontextprotocol/sdk
//
// Run with:
//   npx tsx examples/agents/09-mcp-stdio.ts

import { Agent, mcp } from "../../src/index.js";

// Define the MCP server as a reusable constant
const FilesystemMCP = mcp({
  name: "filesystem",
  transport: "stdio",
  command: "npx",
  args: ["-y", "@modelcontextprotocol/server-filesystem", process.cwd()],
});

const agent = new Agent({
  name: "file-explorer",
  instructions:
    "You are a helpful file explorer. Use the filesystem tools to answer questions about files. Be concise.",
  tools: [FilesystemMCP],
});

const result = await agent.run(
  "List the files in the current directory and tell me what this project is about based on the package.json.",
);

console.log("Answer:", result.output);
console.log("\nTool calls made:");
for (const call of result.meta.toolCalls) {
  console.log(` - ${call.name}(${JSON.stringify(call.input)}) (${call.durationMs}ms)`);
}
console.log("\nIterations:", result.meta.iterations);
console.log("Tokens used:", result.meta.usage.totalTokens);

File: python/examples/agents/09_mcp_stdio.py
# MCP stdio — connect to an MCP server and use its tools
#
# Requires:
#   - the `mcp` Python package (already a dependency of opperai)
#   - npx (ships with Node.js) — used to run the official filesystem MCP server
#
# The official filesystem MCP server is published to npm as
# @modelcontextprotocol/server-filesystem, so we launch it with npx (mirroring
# the TypeScript example).

import asyncio

from opperai.agent import Agent
from opperai.agent.mcp import MCPStdioConfig, mcp


async def main() -> None:
    agent = Agent(
        name="fs-assistant",
        instructions=(
            "You have filesystem access via MCP tools. "
            "Be concise in your responses."
        ),
        tools=[
            mcp(
                MCPStdioConfig(
                    name="filesystem",
                    command="npx",
                    args=["-y", "@modelcontextprotocol/server-filesystem", "/tmp"],
                )
            ),
        ],
    )

    result = await agent.run("List the files in /tmp and tell me how many there are.")
    print("Answer:", result.output)
    print(f"\nIterations: {result.meta.iterations}")
    print(f"Tokens: {result.meta.usage.total_tokens}")
    print(f"Tool calls: {len(result.meta.tool_calls)}")

    # Show which MCP tools were called
    for tc in result.meta.tool_calls:
        print(f"  {tc.name}: {tc.duration_ms:.0f}ms")


if __name__ == "__main__":
    asyncio.run(main())
